File: vidify/player/mpv.py
"""
Mpv is a good alternative to VLC because it's relatively lightweight and
straightforward, but it's mostly used only on Linux systems.

For more information about the player modules, please check out
vidify.player.generic, which contains the abstract base class of which any
player implementation inherits, and an explanation in detail of the methods.
"""

import locale
import logging
from typing import Optional

from mpv import MPV

from vidify.player.generic import PlayerBase
from vidify.config import Config

# Importing locale is necessary since qtpy stomps over the locale settings
# needed by libmpv. This needs to happen after importing PyQT before creating
# the first mpv.MPV instance, so it's in global context.
locale.setlocale(locale.LC_NUMERIC, "C")


class MpvPlayer(PlayerBase):
    # The audio is always muted, which is needed because not all the
    # youtube-dl videos are silent. The keep-open flag stops mpv from closing
    # after the video is over.
    DEFAULT_FLAGS = ["mute"]
    DEFAULT_ARGS = {"vo": "gpu,libmpv,x11", "config": False, "keep-open": "always"}

    def __init__(self, config: Config) -> None:
        # Importing locale is necessary since qtpy stomps over the locale
        # settings needed by libmpv. This needs to happen after importing PyQT
        # before creating the first mpv.MPV instance, so it's in global
        # context.
        locale.setlocale(locale.LC_NUMERIC, "C")

        super().__init__()

        # Converting the flags passed by parameter (str) to a tuple
        # TODO: this won't work from Python
        flags = config.mpv_properties
        flags.extend(self.DEFAULT_FLAGS)

        args = {}
        if config.debug:
            args["log_handler"] = print
            args["loglevel"] = "info"
        args["wid"] = str(int(self.winId()))  # sip.voidptr -> int -> str
        args.update(self.DEFAULT_ARGS)

        self._mpv = MPV(*flags, **args)

    @property
    def pause(self) -> bool:
        return self._mpv.pause

    @pause.setter
    def pause(self, do_pause: bool) -> None:
        logging.info("Playing/Pausing video")
        self._mpv.pause = do_pause

    @property
    def position(self) -> int:
        time = self._mpv.playback_time
        return round(time * 1000) if time is not None else 0

    def seek(self, ms: int, relative: bool = False) -> None:
        """
        Mpv will throw an error if the position is changed before the video
        starts, so this waits for 'seekable' to be set to True.
        """

        self._mpv.wait_for_property("seekable")
        logging.info("Position set to %d milliseconds", ms)
        self._mpv.seek(
            round(ms / 1000, 2), reference="relative" if relative else "absolute"
        )

    def start_video(self, media: str, is_playing: bool = True) -> None:
        logging.info("Started new video")
        self._mpv.play(media)
        # Mpv starts automatically playing the video
        if not is_playing:
            self.pause = True

File: core/player/mpv.rs
//! Mpv is used as the embedded player, which is what it's designed for.
//! It's lightweight, precise and fast, which makes it perfect for
//! synchronizing audio and video with the audiosync extension.

use crate::config::Config;
use crate::error::{Error, Result};
use crate::player::PlayerBase;

use log::info;

type MpvResult<T> = libmpv::Result<T>;

impl From<libmpv::Error> for Error {
    fn from(err: libmpv::Error) -> Self {
        Error::PlayerInit(err.to_string())
    }
}

pub struct Mpv {
    mpv: libmpv::Mpv,
}

impl PlayerBase for Mpv {
    fn new(config: &Config, wid: u64) -> Result<Mpv> {
        // Setting the base properties before mpv is initialized.
        let mpv = libmpv::Mpv::with_initializer(|init| -> MpvResult<()> {
            init.set_property("wid", wid as i64)?;
            init.set_property("vo", "gpu,libmpv,x11")?;
            init.set_property("config", false)?;
            // The audio is always muted, which is needed because not all the
            // youtube-dl videos are silent.
            init.set_property("mute", true)?;
            // The keep-open flag stops mpv from closing after the video is
            // over.
            init.set_property("keep-open", "always")?;

            for (key, val) in config.mpv_properties.iter() {
                init.set_property(key.as_str(), val.as_str())?;
            }

            Ok(())
        })?;

        Ok(Mpv {
            mpv
        })
    }

    fn set_pause(&mut self, do_pause: bool) {
        info!("Setting pause to {}", do_pause);

        // TODO: maybe the trait should return `Result`? This would avoid
        // the unwraps.
        //
        // This doesn't check if it's already paused because mpv already
        // should take care of that.
        // TODO: check that this is true.
        if do_pause {
            self.mpv.pause().unwrap();
        } else {
            self.mpv.unpause().unwrap();
        }
    }

    fn is_paused(&self) -> bool {
        // TODO: maybe the trait should return `Result`? This would avoid
        // the unwrap.
        self.mpv.get_property("pause").unwrap()
    }

    fn position(&self) -> u32 {
        // TODO: maybe the trait should return `Result`? This would avoid
        // the unwrap.
        // TODO: maybe the trait should return u64 or a different type?
        self.mpv.get_property::<i64>("playback_time").unwrap() as u32
    }

    fn seek(&mut self, ms: i64, relative: bool) {
        // TODO: maybe the trait should return `Result`? This would avoid
        // the unwraps.
        // TODO: should this `wait_for_property("seekable")`?
        let secs = (ms as f64 / 1000.0).round();
        if relative {
            info!("Seeking to relative {} ms", ms);
            if secs > 0.0 {
                self.mpv.seek_forward(secs).unwrap();
            } else {
                self.mpv.seek_backward(-secs).unwrap();
            }
        } else {
            info!("Seeking to absolute {} ms", ms);
            self.mpv.seek_absolute(secs).unwrap();
        }
    }

    fn start_video(&mut self, media: &str, start_playing: bool) {
        info!("Starting new media '{}', playing? {}", media, start_playing);

        // TODO: maybe the trait should return `Result`? This would avoid
        // the unwraps.
        self.mpv.command("loadfile", &[media]).unwrap();

        // Mpv starts automatically playing the video.
        if !start_playing {
            self.set_pause(false);
        }
    }
}
